teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,403 @@
+  steps:
+    - container: step-1
+      inputs:
+        - name: source
+          value:
+            - uri: pkg:generic/source
+              digest: sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2
+  steps:
+    - container: step-2
+      outputs:
+        - name: release-file
+          value:
+            - uri: pkg:generic/release-file
+              digest: sha256:33a06c928729e52d1991a2c55765a7c30ef72b098533f220f3f1d6f352fd32e8
+```
+### From Status into Provenance
+
+The Tekton Chains formatter would access the artifact information from the `StepState`. 
+- All `inputs` will be inserted in the `resolvedDependency` section of the `SLSA provenance`.
+- All `outputs` will be placed as `subjects`.
+  - That's because with just information about the `uri` and `digest`, Tekton Chains does not have enough information to distinguish `subjects` from `byProducts`. Infact, even the `StepAction`/`Task`(for inlined-steps) author does not have the context of how an artifact is used (`byProduct` or `subject` of the build). Without some user input, it is not possible for Chains to determine where to store this information so we propose to fit all `output` artifacts as `subjects` in the `SLSA` provenance.
+
+### Transferring Artifacts
+Until we have an Artifact API, when a `Task` or a `StepAction` needs to consume `Artifacts`, it can only do so via `params`. The user of the `Task`/`StepAction` needs to provide it the artifact provenance data. The `Task`/`StepAction` consuming the artifacts will only get the associated metadata which they need to use to fetch the raw data. When passing artifacts between `Tasks/Steps`, only output artifacts of a previous `Task/Step` can be passed as inputs to the consuming `Task/Step`. Because of the structure of [artifact provenance data](#values), we propose passing the artifact provenance data as a json string. **Note** that for large number of artifacts, this string has the potential of getting large enough to cause the `TaskRun` to exceed the `etcd` limit. However, that is true for `artifacts` in general and in the future, we will require some sort of `larger artifacts` mechanism. For, now, this is out of scope.
+
+#### Between Steps
+
+When passing artifacts between `Steps`, we propose the following syntax: `$(steps.<step-name>.outputs.<artifact-name>)` which resolves to the [values](#values) of the associated `output` artifact. If the artifact was not named, the replacement syntax would look like: `$(steps.<step-name>.outputs)`.
+
+`StepActions` can request a `param` to consume artifact provenance data as follows:
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepActions
+metadata:
+  name: step-action-consuming-artifacts
+  description: |
+    This stepaction produces an `output`` artifact called `images`.
+spec:
+  params:
+    - name: image-artifact-provenance
+  image: jq
+  script: |
+    echo $(params.image-artifact-provenance) | jq
```

Comment:
params substitution is blocked by admission webhook

Reply:
AH yes. Direct param substitution in scripts in StepActions is not allowed. We need to set that as an env and pass it that way. Let me send in a fix and I'll ping you after.

Reply:
Fixed!